### PR TITLE
Feature/member task

### PR DESCRIPTION
**Pull request by DzmitryI**

### CODE REVIEW

File: src/components/UI/button/button.js [hunk 1]
```diff
@@ -1,12 +1,20 @@
 import React from 'react';
+import { ThemeContext } from '../../context';
+import { Link } from 'react-router-dom';
 
-const Button = ({ type = 'button', id, name, className, onClick, disabled = false }) => {
+const Button = ({ type = 'button', id, name, className, onClick, disabled = false, to, theme }) => {
   id = id || name;
+  let result = name;
+  if (to) {
+    result = <Link to={to}>{name}</Link>;
```
Comment (maximsan) [inline src/components/UI/button/button.js:9]: let's create a separated component ButtonLink

File: src/components/UI/button/button.js [hunk 2]
```diff
@@ -1,12 +1,20 @@
...
-      {name}
+    <button type={type} className={`${className} ${theme}--btn`} onClick={onClick} disabled={disabled} id={id}>
+      {result}
     </button>
   );
 };
 
-export default Button;
+export default (props) => (
+  <ThemeContext.Consumer>{(theme) => <Button {...props} theme={theme} />}</ThemeContext.Consumer>
```
Comment (maximsan) [inline src/components/UI/button/button.js:19]: let's move it to separated component either

File: src/components/UI/header/header.js [hunk 3]
```diff
@@ -18,25 +21,38 @@ export default class Header extends Component {
...
-        <ul className='nav'>{this.renderLinks(links)}</ul>
-        <ButtonLink className='btn btn-tasks' onClick={this.logoutClick} name='Logout' to={'/Auth'} />
-      </div>
-    );
-  }
-}
+const mapDispatchToProps = (dispatch) => {
+  return {
+    logout: () => dispatch(logout()),
+  };
```
Comment (maximsan) [inline src/components/UI/header/header.js:46]: don't forget also to return dispatch

File: src/components/UI/header/header.js [hunk 4]
```diff
@@ -18,25 +21,38 @@ export default class Header extends Component {
...
+
+export default connect(
+  null,
+  mapDispatchToProps,
+)((props) => (
+  <ThemeContext.Consumer>
+    {(theme) => (
+      <RoleContext.Consumer>{(email) => <Header {...props} email={email} theme={theme} />}</RoleContext.Consumer>
+    )}
+  </ThemeContext.Consumer>
```
Comment (maximsan) [inline src/components/UI/header/header.js:57]: move context upper, you do to difficult

File: src/components/UI/input/input.js [hunk 5]
```diff
@@ -14,12 +14,21 @@ const Input = ({
   onChange,
   errorMessage,
   disabled = false,
+  placeholder = '',
 }) => {
   const result = isInvalid(valid, touched, shouldValidation);
   return (
     <div className={result ? `form-group invalid` : `form-group`}>
```
Comment (maximsan) [inline src/components/UI/input/input.js:21]: shorter

File: src/components/UI/input/input.js [hunk 6]
```diff
@@ -14,12 +14,21 @@ const Input = ({
...
+      <input
+        type={type}
+        id={id}
+        value={value}
+        onChange={onChange}
+        disabled={disabled}
+        placeholder={placeholder}
+        autoComplete={autocomplete}
+      />
       {result ? <span>{errorMessage}</span> : null}
```
Comment (maximsan) [inline src/components/UI/input/input.js:32]: &&

File: src/components/app/App.js [hunk 7]
```diff
@@ -34,52 +32,29 @@ export default class App extends Component {
...
-  };
+  }
 
-  logout = () => {
-    this.setState({ isAuthenticated: false, token: null });
-    localStorage.removeItem('token');
-    localStorage.removeItem('userId');
-    localStorage.removeItem('expirationDate');
-    localStorage.removeItem('email');
+  onColorSwitchClickHandler = (color) => {
```
Comment (maximsan) [inline src/components/app/App.js:54]: this method should be moved also into context

File: src/components/app/App.js [hunk 8]
```diff
@@ -218,7 +181,12 @@ export default class App extends Component {
             <Route
               path='/TasksGrid'
               render={(props) => (
-                <TasksGrid {...props} onCreateTaskClick={this.onCreateTaskClickHandler} isOpen={isTask} />
```
Comment (maximsan) [inline src/components/app/App.js:221]: use component instead of render

File: src/components/app/App.js [hunk 9]
```diff
@@ -242,25 +210,51 @@ export default class App extends Component {
...
+  return {
+    isAuthenticated: !!state.auth.token,
+    email: state.auth.email,
+  };
+};
+
+const mapDispatchToProps = (dispatch) => {
+  return {
+    autoLogin: () => dispatch(autoLogin()),
+  };
```
Comment (maximsan) [inline src/components/app/App.js:257]: don't forget dispatch

File: src/components/auth/auth.js [hunk 10]
```diff
@@ -1,12 +1,12 @@
...
-import axios from 'axios';
-import DisplayNotification from '../../components/displayNotification';
 import { createControl, validateControl } from '../../services/helpers.js';
-import { clearOblectValue } from '../../page/helpersPage';
+import { ToastContainer } from 'react-toastify';
+import { connect } from 'react-redux';
+import { auth } from '../../store/actions/auth';
 
-export default class Auth extends Component {
+class Auth extends Component {
```
Comment (maximsan) [inline src/components/auth/auth.js:9]: PureComponent

File: src/components/colorSwitch/colorSwitch.js [hunk 11]
```diff
@@ -0,0 +1,25 @@
+import React from 'react';
+import imgMoon from '../../assets/images/moon.png';
+import imgSun from '../../assets/images/sun.png';
+
+const colorSwitch = (props) => {
```
Comment (maximsan) [inline src/components/colorSwitch/colorSwitch.js:5]: why is it lowercase ?

File: src/components/displayNotification/displayNotification.js [hunk 12]
```diff
@@ -1,12 +1,22 @@
-const displayNotification = (message) => {
-  const spawnNotification = (title) => {
-    const notification = new Notification(title);
-    setTimeout(notification.close.bind(notification), 4000);
+import React, { Component } from 'react';
+import { ToastContainer, toast } from 'react-toastify';
+import { createNotify } from '../../services/helpers';
+import 'react-toastify/dist/ReactToastify.css';
+
+export default class DisplayNotification extends Component {
```
Comment (maximsan) [inline src/components/displayNotification/displayNotification.js:6]: make it function component

File: src/components/memberTasksGrid/memberTasksGrid.js [hunk 13]
```diff
@@ -111,18 +117,46 @@ export default class MemberTasksGrid extends Component {
...
 }
+
+export default (props) => (
+  <ThemeContext.Consumer>
+    {(theme) => (
+      <RoleContext.Consumer>
+        {(email) => <MemberTasksGrid {...props} theme={theme} email={email} />}
+      </RoleContext.Consumer>
+    )}
+  </ThemeContext.Consumer>
```
Comment (maximsan) [inline src/components/memberTasksGrid/memberTasksGrid.js:161]: move outside

File: src/components/spinner/spinner.js [hunk 14]
```diff
@@ -1,8 +1,9 @@
 import React from 'react';
+import { ThemeContext } from '../context';
 
-const spinner = () => {
+const Spinner = (props) => {
```
Comment (maximsan) [inline src/components/spinner/spinner.js:4]: spread props

File: src/components/spinner/spinner.js [hunk 15]
```diff
@@ -11,4 +12,6 @@ const spinner = () => {
   );
 };
 
-export default spinner;
+export default (props) => (
+  <ThemeContext.Consumer>{(theme) => <Spinner {...props} theme={theme} />}</ThemeContext.Consumer>
```
Comment (maximsan) [inline src/components/spinner/spinner.js:16]: move outside

File: src/components/taskTracksGrid/taskTracksGrid.js [hunk 16]
```diff
@@ -47,22 +50,25 @@ export default class TaskTracsGrid extends Component {
...
     }
     const tracks = await updateMemberProgress('', this.state.taskId);
     this.setState({ tracks, loading: false });
   };
 
   render() {
     const { tracks, loading } = this.state;
+    const { email, theme } = this.props;
+    const admin = TABLE_ROLES.ADMIN;
+    const mentor = TABLE_ROLES.MENTOR;
```
Comment (maximsan) [inline src/components/taskTracksGrid/taskTracksGrid.js:63]: these variables are redundant

File: src/hoc/layout.js [hunk 17]
```diff
@@ -1,11 +1,7 @@
 import React from 'react';
 
 const layout = (props) => {
-  return (
-    <div>
-      <main>{props.children}</main>
-    </div>
-  );
+  return <div>{props.children}</div>;
```
Comment (maximsan) [inline src/hoc/layout.js:4]: what it the sense of the component if it's empty?

File: src/components/spinner/spinner.js (hunk 14 above)
Comment (DzmitryI) [reply src/components/spinner/spinner.js:4]: fixed

File: src/components/spinner/spinner.js (hunk 15 above)
Comment (DzmitryI) [reply src/components/spinner/spinner.js:16]: fixed

File: src/components/taskTracksGrid/taskTracksGrid.js (hunk 16 above)
Comment (DzmitryI) [reply src/components/taskTracksGrid/taskTracksGrid.js:63]: fixed

File: src/components/UI/button/button.js (hunk 2 above)
Comment (DzmitryI) [reply src/components/UI/button/button.js:19]: fixed

File: src/components/UI/header/header.js (hunk 4 above)
Comment (DzmitryI) [reply src/components/UI/header/header.js:57]: fixed

File: src/components/colorSwitch/colorSwitch.js (hunk 11 above)
Comment (DzmitryI) [reply src/components/colorSwitch/colorSwitch.js:5]: fixed

File: src/components/memberTasksGrid/memberTasksGrid.js (hunk 13 above)
Comment (DzmitryI) [reply src/components/memberTasksGrid/memberTasksGrid.js:161]: fixed

File: src/components/auth/auth.js (hunk 10 above)
Comment (DzmitryI) [reply src/components/auth/auth.js:9]: fixed

File: src/components/UI/button/button.js (hunk 1 above)
Comment (DzmitryI) [reply src/components/UI/button/button.js:9]: fixed

File: src/components/UI/input/input.js (hunk 5 above)
Comment (DzmitryI) [reply src/components/UI/input/input.js:21]: fixed

File: src/components/UI/input/input.js (hunk 6 above)
Comment (DzmitryI) [reply src/components/UI/input/input.js:32]: fixed

File: src/components/app/App.js (hunk 7 above)
Comment (DzmitryI) [reply src/components/app/App.js:54]: fixed

File: src/components/displayNotification/displayNotification.js (hunk 12 above)
Comment (DzmitryI) [reply src/components/displayNotification/displayNotification.js:6]: in this component I need the method ComponentDitMount()

**Pull request merged**